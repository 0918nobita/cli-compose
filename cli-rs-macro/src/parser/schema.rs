use std::fmt;

use syn::{
    parse::{Parse, ParseStream},
    Ident, Token,
};

use super::{field_schema::FieldSchema, modifiers::Modifiers};
use crate::parser::{arg_kind::ArgKind, modifier::Modifier};

pub struct Schema {
    pub kind: ArgKind,
    pub modifiers: Modifiers,
    pub data: SchemaData,
}

impl fmt::Debug for Schema {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "Schema({:?}, {:?}, {:?})",
            self.kind, self.modifiers, self.data
        )
    }
}

#[derive(Debug)]
pub enum SchemaData {
    Single(FieldSchema),
    Multiple(Vec<FieldSchema>),
}

impl Parse for Schema {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let ident = input.parse::<Ident>()?;

        let kind = ArgKind::try_from(ident)?;

        let modifiers = if input.peek(syn::token::Paren) {
            let parenthesized;
            syn::parenthesized!(parenthesized in input);

            let modifiers = parenthesized
                .parse_terminated::<Modifier, Token![,]>(Modifier::parse)?
                .into_iter()
                .collect::<Vec<_>>();

            match Modifiers::try_from_slice(modifiers.as_slice()) {
                Ok(modifiers) => modifiers,
                Err(modifier) => {
                    return Err(syn::Error::new_spanned(modifier.name, "Duplicate modifier"))
                }
            }
        } else {
            Modifiers::default()
        };

        let data = if input.peek(syn::token::Brace) {
            let braced;
            syn::braced!(braced in input);

            let field_schemas = braced
                .parse_terminated::<FieldSchema, Token![,]>(FieldSchema::parse)?
                .into_iter()
                .collect::<Vec<_>>();

            SchemaData::Multiple(field_schemas)
        } else {
            let field_schema = input.parse::<FieldSchema>()?;
            SchemaData::Single(field_schema)
        };

        Ok(Self {
            kind,
            modifiers,
            data,
        })
    }
}
