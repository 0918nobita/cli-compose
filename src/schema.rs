use derive_more::Display;

pub use cli_compose_macro::{ArgOpt, Cli, FromKebabStr, MultiSelect, Opt, PosArg, SingleSelect};

#[derive(Display)]
pub enum Flag {
    #[display(fmt = "--{}", _0)]
    LongOnly(String),

    #[display(fmt = "-{}", _0)]
    ShortOnly(char),

    #[display(fmt = "--{}, -{}", _0, _1)]
    BothLongAndShort(String, char),
}

pub trait AsPosArg: Sized {
    fn name() -> String;

    fn description() -> String;

    fn parse(s: &str) -> Option<Self>;
}

pub trait AsArgOpt: Sized {
    fn flag() -> Flag;

    fn description() -> String;

    fn parse(s: &str) -> Option<Self>;
}

pub trait AsOpt {
    fn flag() -> Flag;

    fn description() -> String;
}

pub trait AsSingleSelect {
    fn name() -> String;

    fn description() -> String;
}

pub trait AsMultiSelect {
    fn name() -> String;

    fn description() -> String;
}
