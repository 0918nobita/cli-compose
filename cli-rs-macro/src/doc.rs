use syn::{Attribute, Lit, Meta};

fn try_get_single_line_doc(attr: &Attribute) -> Option<String> {
    let meta = attr.parse_meta().ok()?;

    let lit_str = match meta {
        Meta::NameValue(syn::MetaNameValue {
            path,
            lit: Lit::Str(lit_str),
            ..
        }) if path.is_ident("doc") => lit_str,
        _ => return None,
    };

    Some(lit_str.value().trim_start().to_owned())
}

pub fn extract_doc(attrs: &[Attribute]) -> String {
    attrs
        .iter()
        .filter_map(try_get_single_line_doc)
        .collect::<Vec<_>>()
        .join("\n")
}
