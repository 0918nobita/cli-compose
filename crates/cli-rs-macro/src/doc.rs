use syn::Attribute;

fn try_get_single_line_doc(attr: Attribute) -> Option<String> {
    let meta = attr.parse_meta().ok()?;
    match meta {
        syn::Meta::NameValue(syn::MetaNameValue {
            path,
            lit: syn::Lit::Str(lit_str),
            ..
        }) if path.is_ident("doc") => Some(lit_str.value().trim_start().to_owned()),
        _ => None,
    }
}

pub fn extract_doc(attrs: impl Iterator<Item = Attribute>) -> String {
    attrs
        .filter_map(try_get_single_line_doc)
        .collect::<Vec<_>>()
        .join("\n")
}
