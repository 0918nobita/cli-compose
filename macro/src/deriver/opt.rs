use bae::FromAttributes;
use convert_case::{Case, Casing};
use proc_macro2::TokenStream;
use quote::quote;

use crate::doc::extract_doc;

#[derive(FromAttributes)]
struct Opt {
    long: Option<syn::LitStr>,

    short: Option<syn::LitChar>,

    #[allow(dead_code)]
    short_only: Option<()>,
}

pub fn derive_opt(input: TokenStream) -> syn::Result<TokenStream> {
    let input = syn::parse2::<syn::DeriveInput>(input)?;

    let attr = Opt::try_from_attributes(&input.attrs)?;

    let struct_name = &input.ident;
    let long = attr
        .as_ref()
        .and_then(|attr| attr.long.clone())
        .map_or_else(
            || struct_name.to_string().to_case(Case::Kebab),
            |lit_str| lit_str.value(),
        );

    let flag = match &attr.and_then(|opt| opt.short) {
        Some(short) => {
            quote! { cli_compose::schema::Flag::BothLongAndShort(#long.to_owned(), #short) }
        }
        None => quote! { cli_compose::schema::Flag::LongOnly(#long.to_owned()) },
    };

    let doc = extract_doc(&input.attrs);

    let sharp = syn::Token![#](proc_macro2::Span::call_site());

    Ok(quote! {
        impl cli_compose::schema::AsMember for #struct_name {
            fn handle(mut builder: cli_compose::schema::CliBuilder) -> cli_compose::schema::CliBuilder {
                use cli_compose::schema::{forwarded::quote::quote, AsOpt};

                let flag = format!("{}", <#struct_name as AsOpt>::flag());

                builder.ops.extend(quote! {
                    println!("ArgOpt {}", #sharp flag);
                });

                builder
            }
        }

        impl cli_compose::schema::AsOpt for #struct_name {
            fn flag() -> cli_compose::schema::Flag {
                #flag
            }

            fn description() -> String {
                #doc.to_owned()
            }
        }
    })
}

#[cfg(test)]
mod tests {
    use quote::quote;

    fn test_opt_deriver(input: proc_macro2::TokenStream) -> anyhow::Result<String> {
        let tokens = super::derive_opt(input)?;

        crate::pretty_print::pretty_print_rust_code(tokens)
    }

    #[test]
    fn empty() {
        insta::assert_debug_snapshot!(test_opt_deriver(quote! {}));
    }
}
