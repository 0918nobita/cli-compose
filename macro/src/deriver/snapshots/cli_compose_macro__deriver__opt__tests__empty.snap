---
source: macro/src/deriver/opt.rs
assertion_line: 82
expression: "test_opt_deriver(quote! {})"
---
Err(
    Error(
        "unexpected end of input, expected one of: `struct`, `enum`, `union`",
    ),
)
